drivers/pinctrl/pinctrl_iopctl.c

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nxp_iopctl
+
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/init.h>
+
+#define OFFSET(mux) (((mux) & 0xFFF00000) >> 20)
+#define INDEX(mux)  (((mux) & 0xF0000) >> 16)
+
+/* Port register addresses. */
+static uint32_t *iopctl[] = {
+	(uint32_t *)DT_REG_ADDR(DT_NODELABEL(iopctl0)),
+#if DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) > 1
+	(uint32_t *)DT_REG_ADDR(DT_NODELABEL(iopctl1)),
+#endif
+#if DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) > 2
+	(uint32_t *)DT_REG_ADDR(DT_NODELABEL(iopctl2)),
+#endif
+};
+
+int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt, uintptr_t reg)
+{
+	for (uint8_t i = 0; i < pin_cnt; i++) {
+		uint32_t pin_mux = pins[i];
+		uint32_t index = INDEX(pin_mux);
+		uint32_t offset = OFFSET(pin_mux);
+		/* Set pinmux */
+		*(iopctl[index] + offset) = pin_mux;
```

Comment:
if in this case, why we need hal update?

Reply:
Because RT500/600 pin header files had been updated to add one parameter to represent instance index. There are three iopctl instances on RT700, and current pin model can't adapt to RT700 with a unified model. Aligned with Daniel, he approve to create a new model for RT 3 digital platforms.